### ISSUE TITLE
App router slow performance when using dynamic route

### ISSUE THREAD

**Issue by manuelseisl** [vercel/next.js#50332]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.3.0: Mon Jan 30 20:38:43 PST 2023; root:xnu-8792.81.3~2/RELEASE_ARM64_T8112
    Binaries:
      Node: 18.16.0
      npm: 9.6.6
    Relevant packages:
      next: 13.4.3
      eslint-config-next: 13.4.3
      react: 18.2.0
      react-dom: 18.2.0
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

Routing (next/router, next/navigation, next/link)

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/manuelseisl/nextjs-demo

### To Reproduce

**Open link:**
https://nextjs-demo-six-kohl.vercel.app

**Go to "console -> network"**
Document loading time is about 30 - 40 ms

**when opening link with dynamic route:**
https://nextjs-demo-six-kohl.vercel.app/de/asdf
Document loading time is about 170 - 180 ms

### Describe the Bug

Loading performance is about 4 times slower when using dynamic routes. We also tried with page router and the result was the expected behavior

### Expected Behavior

Better loading performance for document when using dynamic routes with app router

### Which browser are you using? (if relevant)

Chrome 113.0.5672.127

### How are you deploying your application? (if relevant)

Vercel

**Comment by altechzilla** [comment]
Similar issue here.

The app router is incredibly slow for me during dev as well. Also it looks like it prefetches routes that it shouldn't, including hundreds of dynamic paths - flooding the server memory pretty much. 7000 modules that used to be compiled in 6 seconds using the /pages directory now take somewhere like 60-90 seconds. And that's per route.

Not sure if it's related to appDir (which I assume it is) or that it doesn't config.modularizeImports properly?

**Comment by joacub** [comment]
> Similar issue here.
> 
> The app router is incredibly slow for me during dev as well. Also it looks like it prefetches routes that it shouldn't, including hundreds of dynamic paths - flooding the server memory pretty much. 7000 modules that used to be compiled in 6 seconds using the /pages directory now take somewhere like 60-90 seconds. And that's per route.
> 
> Not sure if it's related to appDir (which I assume it is) or that it doesn't config.modularizeImports properly?

Same 7000 modules takes 40 seconds

**Comment by poorscousertommy8** [comment]
I can replicate that too. The root path is fine, dynamic paths are actually quite a bit slower. This phenomenon **did not** occur in **pages router**.

**Comment by altechzilla** [comment]
Sometimes it takes even 2-3 minutes for the initial build. And when navigating between routes or when editing a file, it's like it re-compiles all modules. Maybe there's a bug between the App directory, transpiled modules, and modularized imports? It's acting as if no modules are getting modularized (i.e. @mui) and as if it needs to reload all of them for some reason.

**Comment by altechzilla** [comment]
![image](https://github.com/vercel/next.js/assets/8274669/45da3ee2-ddcf-452b-932d-42ce980d0514)

Running this on an 8-Core Intel Core i9 64 GB RAM so it's not a hardware resource issue. In pages dir it's around 6 seconds. But in appDir, RIP.

**Comment by iukea1** [comment]
Have any of you guys tried this with turbo pack?

(I have not used it nor known its limitations)

**Comment by joacub** [comment]
> Have any of you guys tried this with turbo pack?
> 
> (I have not used it nor known its limitations)

Turbopack is even worse getting 50 seconds to build simple pages. Hot reloading is fast super fast but cold or lazy build is a disaster

**Comment by iukea1** [comment]
> > Have any of you guys tried this with turbo pack?
> > 
> > (I have not used it nor known its limitations)
> 
> Turbopack is even worse getting 50 seconds to build simple pages. Hot reloading is fast super fast but cold or lazy build is a disaster 

Yaaaa damnnnn 

I saw another thread about 1 hour ago where people were talking about how slow turbopack is.

**Comment by jilv220** [comment]
> Similar issue here.
> 
> The app router is incredibly slow for me during dev as well. Also it looks like it prefetches routes that it shouldn't, including hundreds of dynamic paths - flooding the server memory pretty much. 7000 modules that used to be compiled in 6 seconds using the /pages directory now take somewhere like 60-90 seconds. And that's per route.
> 
> Not sure if it's related to appDir (which I assume it is) or that it doesn't config.modularizeImports properly?

Yep, it prefetches too much for some reason.

**Comment by altechzilla** [comment]
Turbopack won't work me at all. It can't resolve paths and it won't even start up. But why would the app router have a 20X worse performance?

**Comment by duongductrong** [comment]
The same, hope this topic can resolve soon :((

**Comment by altechzilla** [comment]
Even on an M2 Max with 96GB RAM it still takes an insane amount of time to compile. [Someone here](https://github.com/vercel/next.js/issues/50361#issuecomment-1567045579) suggested a downgrade to an older Next version, but there's gotta be a tweak coming, right? I mean this is a pretty critical bug in Next.

**Comment by joacub** [comment]
> Even on an M2 Max with 96GB RAM it still takes an insane amount of time to compile. [Someone here](https://github.com/vercel/next.js/issues/50361#issuecomment-1567045579) suggested a downgrade to an older Next version, but there's gotta be a tweak coming, right? I mean this is a pretty critical bug in Next.

Seems like they are working on it, there is one pending update of turbopack and swc that seems promising get faster compiling times. So let’s see in a couple hours when they release the next canary version.

**Comment by joacub** [comment]
Check this ———> https://github.com/vercel/next.js/pull/50519

**Comment by poorscousertommy8** [comment]
originally it was about the performance in production mode. the catch-all routes are slower in response than in the pre App Router version. Hopefully this will be fixed soon.

**Comment by altechzilla** [comment]
> Check this ———> #50519

Downgraded to 13.2.3 and compile times are a LOT better. I'm not using Turbopack because of the complexity of our project and all the unsupported features, but I think webpack projects should get a fix too. Also referenced a bunch of other related issues because this seems to be related to something other than just dynamic routes. It's pretty much every route.

<img width="879" alt="image" src="https://github.com/vercel/next.js/assets/8274669/5df08701-b3e7-4589-9224-4022a089163f">

**Comment by alexdanilowicz** [comment]
+1 to all of this. `_app` router slowness is unacceptable, will be downgrading for now.

```
- event compiled client and server successfully in 8.7s (4597 modules)
```

**Comment by joacub** [comment]
> Check this ———> #50519

No change, I think now is even worse…. This is stating to be really annoying, and now I’m starting to think that no body cares at nextjs team…, this needs to be resolved very soon this can no go longer, this way is impossible to work with 40~ second compiling times, does not matter if Turbopack or webpack, webpack is even faster ….. I will appreciate any response from next team to know what is the status of this and or if has solution which I think, of course have. I may go back (if this stay this way longer) to my own implementation which was way way faster without compiling enter you the whole project in 30 seconds and then no wait per lazy builds and super fast hmr.

**Comment by altechzilla** [comment]
> > Check this ———> #50519
> 
> No change, I think now is even worse…. This is stating to be really annoying, and now I’m starting to think that no body cares at nextjs team…, this needs to be resolved very soon this can no go longer, this way is impossible to work with 40~ second compiling times, does not matter if Turbopack or webpack, webpack is even faster ….. I will appreciate any response from next team to know what is the status of this and or if has solution which I think, of course have. I may go back (if this stay this way longer) to my own implementation (no nextjs) which was way way faster without compiling enter you the whole project in 30 seconds and then no wait per lazy builds and super fast hmr.

Are you using @mui/icons-material or another large icons library? I started changing all tree shaked imports (like `import { DeleteOutlineOutlined } from '@mui/icons-material';` to `import DeleteOutlineOutlined from '@mui/icons-material/DeleteOutlineOutlined';` - even though the library is using the `modularizeImports` option in the Next config, and I see SOME degree of improvement, but still slow. I had to get a M2 Max 96GB RAM MacBook to reduce the compile times. But obviously some random memory leak or whatever shouldn't be a reason to force the need for expensive hardware.

**Comment by joacub** [comment]
> > > Check this ———> #50519
> > 
> > 
> > No change, I think now is even worse…. This is stating to be really annoying, and now I’m starting to think that no body cares at nextjs team…, this needs to be resolved very soon this can no go longer, this way is impossible to work with 40~ second compiling times, does not matter if Turbopack or webpack, webpack is even faster ….. I will appreciate any response from next team to know what is the status of this and or if has solution which I think, of course have. I may go back (if this stay this way longer) to my own implementation (no nextjs) which was way way faster without compiling enter you the whole project in 30 seconds and then no wait per lazy builds and super fast hmr.
> 
> Are you using @mui/icons-material or another large icons library? I started changing all tree shaked imports (like `import { DeleteOutlineOutlined } from '@mui/icons-material';` to `import DeleteOutlineOutlined from '@mui/icons-material/DeleteOutlineOutlined';` - even though the library is using the `modularizeImports` option in the Next config, and I see SOME degree of improvement, but still slow. I had to get a more powerful hardware to reduce the compile times. But obviously some random memory leak or whatever shouldn't be a reason to force the need for expensive hardware.

I’m doing the same, and it’s slow as a hell, ranging from 10-50 seconds build

**Comment by JunkyDeLuxe** [comment]
There is a hudge bug or something is not correct in the implentation of the App router. I use the last canary version v13.4.5 and i can confirm, with next dev (under docker), the app router is very very slow, I have to wait sometimes some seconds, almost ten (WTF), in order to display a simple web page with dynamic fetch GET method, with some client components in it ... Everything is compiling a lot of time, a lot of modules are compiling, 3000+, 4000+, dunno what is going here
Why there is no response from the NextJS core team about that ...

**Comment by joshua-lawrence** [comment]
Also having a same issue, starting to regret using Next.js for development because of the horrible performance in the dev environment.

**Comment by joacub** [comment]
There are several commits these days about performance, and seems working worse,  … going in the wrong direction

**Comment by timneutkens** [comment]
I've posted a reply there. Let's consolidate these issues into one as they're all about the same thing.
https://github.com/vercel/next.js/issues/48748#issuecomment-1578374105

> There are several commits these days about performance, and seems working worse, … going in the wrong direction

I understand that you're frustrated. In order for me to help please provide the `.next/trace` as outlined here: https://github.com/vercel/next.js/issues/48748#issuecomment-1578374105. It will allow me to provide you with insights into why your application is compiling unexpectedly slow. Thanks Johan 🙏

**Comment by joacub** [comment]
> I've posted a reply there. Let's consolidate these issues into one as they're all about the same thing. [#48748 (comment)](https://github.com/vercel/next.js/issues/48748#issuecomment-1578374105)
> 
> > There are several commits these days about performance, and seems working worse, … going in the wrong direction
> 
> I understand that you're frustrated. In order for me to help please provide the `.next/trace` as outlined here: [#48748 (comment)](https://github.com/vercel/next.js/issues/48748#issuecomment-1578374105). It will allow me to provide you with insights into why your application is compiling unexpectedly slow. Thanks Johan 🙏

I will when have a chance send you a trace, and for turbopack I don’t know how to send you that as turbo pack does not yield any trace.

But you can simulate this very easily using material ui with next-intl and you will see this only putting one component. I have optimized already the import as much as I can and still really slow, 30 seconds per page.

thanks for responding.

**Comment by Gm33** [comment]
any news on this? Problem still occurs

**Comment by rusted-love** [comment]
The App directory was a great idea, but the realization is very disappointing. We encountered more than 16 bugs that made our development and production much worse than before.

**Comment by timneutkens** [comment]
@Gm33 please follow the other issue I linked and provide the trace + cpuprofile:  https://github.com/vercel/next.js/issues/48748#issuecomment-1578374105

Besides that we shipped 13.4.8 yesterday that holds many development performance optimizations: https://twitter.com/nextjs/status/1675929873720680459

> We encountered more than 16 bugs that made our development and production much worse than before.

@rusted-love Sorry to hear that you're struggling, we're working hard on fixing bugs and optimizing performance.
I had a look at the bug reports and it seems you've opened 2 issues (both closed) and commented on 7 issues (including your own). Can you report the other issues you've run into so that we can triage them? Thank you!
